Compile error

OS: Macos 
Arch: arm64(Darwin)
Steps to reproduce:
1) Clone zld
2) With the latest binary (dev.1836+28364166e)
```
zig_repo/zig build  
/_path_/zld/build.zig:27:8: error: no field or member function named 'addPackagePath' in 'Build.CompileStep'
    lib.addPackagePath("dis_x86_64", "zig-dis-x86_64/src/dis_x86_64.zig");
    ~~~^~~~~~~~~~~~~~~
/path_to_zig_repo/lib/std/Build/CompileStep.zig:1:1: note: struct declared here
const builtin = @import("builtin");
```
Earlier binaries produce other compile errors.

--- a/build.zig
+++ b/build.zig
@@ -17,11 +17,9 @@ pub fn build(b: *Builder) void {
     const is_qemu_enabled = b.option(bool, "enable-qemu", "Use QEMU to run cross compiled foreign architecture tests") orelse false;
     const enable_tracy = b.option([]const u8, "tracy", "Enable Tracy integration. Supply path to Tracy source");
 
-    b.addModule(.{
-        .name = "dis_x86_64",
+    const dis_x86_64 = b.addModule("dis_x86_64", .{
         .source_file = .{ .path = "zig-dis-x86_64/src/dis_x86_64.zig" },
     });
-    const dis_x86_64 = b.modules.get("dis_x86_64").?;
 
     const exe = b.addExecutable(.{
         .name = "zld",
@@ -76,11 +74,8 @@ pub fn build(b: *Builder) void {
         .root_source_file = .{ .path = "src/test.zig" },
         .optimize = mode,
     });
-    const test_base = b.createModule(.{ .source_file = .{ .path = "src/test.zig" } });
-    const e2e_tests = b.createModule(.{ .source_file = .{ .path = "test/test.zig" } });
-    e2e_tests.dependencies.put("test_base", test_base) catch @panic("OOM");
-    tests.addModule("end_to_end_tests", e2e_tests);
     tests.addModule("dis_x86_64", dis_x86_64);
+    tests.main_pkg_path = "."; // set root directory as main package path for our tests
 
     const test_opts = b.addOptions();
     tests.addOptions("build_options", test_opts);

--- a/src/Elf.zig
+++ b/src/Elf.zig
@@ -215,7 +215,7 @@ pub fn flush(self: *Elf) !void {
     try self.parsePositionals(positionals.items);
     try self.parseLibs(libs.keys());
 
-    for (self.objects.items) |_, object_id| {
+    for (self.objects.items, 0..) |_, object_id| {
         try self.resolveSymbolsInObject(@intCast(u16, object_id));
     }
     try self.resolveSymbolsInArchives();
@@ -236,7 +236,7 @@ pub fn flush(self: *Elf) !void {
         try object.scanInputSections(self);
     }
 
-    for (self.objects.items) |*object, object_id| {
+    for (self.objects.items, 0..) |*object, object_id| {
         try object.splitIntoAtoms(self.base.allocator, @intCast(u16, object_id), self);
     }
 
@@ -517,7 +517,7 @@ fn insertSection(self: *Elf, shdr: elf.Elf64_Shdr, shdr_name: []const u8) !u16 {
     // allocated within each respective segment. Of course, it is good practice to have
     // the sections sorted, but it's a useful hack we can use for the debug builds in
     // self-hosted Zig compiler.
-    const insertion_index = for (self.sections.items(.shdr)) |oshdr, i| {
+    const insertion_index = for (self.sections.items(.shdr), 0..) |oshdr, i| {
         const oshdr_name = self.shstrtab.getAssumeExists(oshdr.sh_name);
         if (getSectionPrecedence(oshdr, oshdr_name) > precedence) break @intCast(u16, i);
     } else @intCast(u16, self.sections.items(.shdr).len);
@@ -725,7 +725,7 @@ fn resolveSymbolsInObject(self: *Elf, object_id: u16) !void {
 
     log.debug("resolving symbols in {s}", .{object.name});
 
-    for (object.symtab.items) |sym, i| {
+    for (object.symtab.items, 0..) |sym, i| {
         const sym_id = @intCast(u32, i);
         const sym_name = self.getSymbolName(.{ .sym_index = sym_id, .file = object_id });
         const st_bind = sym.st_info >> 4;
@@ -1115,7 +1115,7 @@ fn allocateNonAllocSections(self: *Elf) !void {
 
 fn allocateAtoms(self: *Elf) !void {
     const slice = self.sections.slice();
-    for (slice.items(.last_atom)) |last_atom, i| {
+    for (slice.items(.last_atom), 0..) |last_atom, i| {
         var atom = last_atom orelse continue;
         const shdr_ndx = @intCast(u16, i);
         const shdr = slice.items(.shdr)[shdr_ndx];
@@ -1194,7 +1194,7 @@ pub fn logAtom(self: *Elf, atom: *const Atom, comptime logger: anytype) void {
 
 fn logAtoms(self: *Elf) void {
     const slice = self.sections.slice();
-    for (slice.items(.last_atom)) |last_atom, i| {
+    for (slice.items(.last_atom), 0..) |last_atom, i| {
         var atom = last_atom orelse continue;
         const ndx = @intCast(u16, i);
         const shdr = slice.items(.shdr)[ndx];
@@ -1216,7 +1216,7 @@ fn logAtoms(self: *Elf) void {
 
 fn writeAtoms(self: *Elf) !void {
     const slice = self.sections.slice();
-    for (slice.items(.last_atom)) |last_atom, i| {
+    for (slice.items(.last_atom), 0..) |last_atom, i| {
         var atom = last_atom orelse continue;
         const shdr_ndx = @intCast(u16, i);
         const shdr = slice.items(.shdr)[shdr_ndx];
@@ -1301,7 +1301,7 @@ fn writeSymtab(self: *Elf) !void {
     });
 
     for (self.objects.items) |object| {
-        for (object.symtab.items) |sym, sym_id| {
+        for (object.symtab.items, 0..) |sym, sym_id| {
             if (sym.st_name == 0) continue;
             const st_bind = sym.st_info >> 4;
             const st_type = sym.st_info & 0xf;
@@ -1421,7 +1421,7 @@ fn writeHeader(self: *Elf) !void {
 }
 
 pub fn getSectionByName(self: *Elf, name: []const u8) ?u16 {
-    for (self.sections.items(.shdr)) |shdr, i| {
+    for (self.sections.items(.shdr), 0..) |shdr, i| {
         const this_name = self.shstrtab.getAssumeExists(shdr.sh_name);
         if (mem.eql(u8, this_name, name)) return @intCast(u16, i);
     } else return null;
@@ -1478,7 +1478,7 @@ pub fn getEntryPoint(self: Elf) error{EntrypointNotFound}!SymbolWithLoc {
 
 fn logSections(self: Elf) void {
     log.debug("sections:", .{});
-    for (self.sections.items(.shdr)) |shdr, i| {
+    for (self.sections.items(.shdr), 0..) |shdr, i| {
         log.debug("  sect({d}): {s} @{x}, sizeof({x})", .{
             i,
             self.shstrtab.getAssumeExists(shdr.sh_name),
@@ -1491,7 +1491,7 @@ fn logSections(self: Elf) void {
 fn logSymtab(self: Elf) void {
     for (self.objects.items) |object| {
         log.debug("locals in {s}", .{object.name});
-        for (object.symtab.items) |sym, i| {
+        for (object.symtab.items, 0..) |sym, i| {
             // const st_type = sym.st_info & 0xf;
             const st_bind = sym.st_info >> 4;
             // if (st_bind != elf.STB_LOCAL or st_type != elf.STT_SECTION) continue;

--- a/src/Elf/Object.zig
+++ b/src/Elf/Object.zig
@@ -83,7 +83,7 @@ pub fn parse(self: *Object, allocator: Allocator, cpu_arch: std.Target.Cpu.Arch)
 
     if (self.header.e_shnum == 0) return;
 
-    for (self.getShdrs()) |shdr, i| switch (shdr.sh_type) {
+    for (self.getShdrs(), 0..) |shdr, i| switch (shdr.sh_type) {
         elf.SHT_SYMTAB => {
             self.symtab_index = @intCast(u16, i);
             const nsyms = @divExact(shdr.sh_size, @sizeOf(elf.Elf64_Sym));
@@ -137,28 +137,28 @@ pub fn splitIntoAtoms(self: *Object, allocator: Allocator, object_id: u16, elf_f
     var rel_shdrs = std.AutoHashMap(u16, u16).init(allocator);
     defer rel_shdrs.deinit();
 
-    for (shdrs) |shdr, i| switch (shdr.sh_type) {
+    for (shdrs, 0..) |shdr, i| switch (shdr.sh_type) {
         elf.SHT_REL, elf.SHT_RELA => {
             try rel_shdrs.putNoClobber(@intCast(u16, shdr.sh_info), @intCast(u16, i));
         },
         else => {},
     };
 
-    for (shdrs) |shdr, i| switch (shdr.sh_type) {
+    for (shdrs, 0..) |shdr, i| switch (shdr.sh_type) {
         elf.SHT_PROGBITS, elf.SHT_NOBITS => {
             try symbols_by_shndx.putNoClobber(@intCast(u16, i), std.ArrayList(u32).init(allocator));
         },
         else => {},
     };
 
-    for (self.getSourceSymtab()) |sym, sym_id| {
+    for (self.getSourceSymtab(), 0..) |sym, sym_id| {
         if (sym.st_shndx == elf.SHN_UNDEF) continue;
         if (elf.SHN_LORESERVE <= sym.st_shndx and sym.st_shndx < elf.SHN_HIRESERVE) continue;
         const map = symbols_by_shndx.getPtr(sym.st_shndx) orelse continue;
         try map.append(@intCast(u32, sym_id));
     }
 
-    for (shdrs) |shdr, i| switch (shdr.sh_type) {
+    for (shdrs, 0..) |shdr, i| switch (shdr.sh_type) {
         elf.SHT_PROGBITS, elf.SHT_NOBITS => {
             const ndx = @intCast(u16, i);
             const shdr_name = self.getShString(shdr.sh_name);

--- a/src/MachO.zig
+++ b/src/MachO.zig
@@ -857,7 +857,7 @@ fn parseLibs(
     const tracy = trace(@src());
     defer tracy.end();
 
-    for (lib_names) |lib, i| {
+    for (lib_names, 0..) |lib, i| {
         const lib_info = lib_infos[i];
         log.debug("parsing lib path '{s}'", .{lib});
         if (try self.parseDylib(lib, dependent_libs, .{
@@ -1476,7 +1476,7 @@ fn resolveSymbols(self: *MachO, resolver: *SymbolResolver) !void {
     const tracy = trace(@src());
     defer tracy.end();
 
-    for (self.objects.items) |_, object_id| {
+    for (self.objects.items, 0..) |_, object_id| {
         try self.resolveSymbolsInObject(@intCast(u16, object_id), resolver);
     }
 
@@ -1640,7 +1640,7 @@ fn resolveSymbolsInDylibs(self: *MachO, resolver: *SymbolResolver) !void {
         const sym = self.getSymbolPtr(global);
         const sym_name = self.getSymbolName(global);
 
-        for (self.dylibs.items) |dylib, id| {
+        for (self.dylibs.items, 0..) |dylib, id| {
             if (!dylib.symbols.contains(sym_name)) continue;
 
             const dylib_id = @intCast(u16, id);
@@ -1767,7 +1767,7 @@ fn resolveDyldStubBinder(self: *MachO, resolver: *SymbolResolver) !void {
     const global = SymbolWithLoc{ .sym_index = sym_index };
     try self.globals.append(gpa, global);
 
-    for (self.dylibs.items) |dylib, id| {
+    for (self.dylibs.items, 0..) |dylib, id| {
         if (!dylib.symbols.contains(sym_name)) continue;
 
         const dylib_id = @intCast(u16, id);
@@ -1861,7 +1861,7 @@ fn createSegments(self: *MachO) !void {
         });
     }
 
-    for (self.sections.items(.header)) |header, sect_id| {
+    for (self.sections.items(.header), 0..) |header, sect_id| {
         if (header.size == 0) continue; // empty section
 
         const segname = header.segName();
@@ -1935,7 +1935,7 @@ fn splitIntoAtoms(self: *MachO) !void {
     const tracy = trace(@src());
     defer tracy.end();
 
-    for (self.objects.items) |*object, object_id| {
+    for (self.objects.items, 0..) |*object, object_id| {
         try object.splitIntoAtoms(self, @intCast(u31, object_id));
     }
 }
@@ -1947,7 +1947,7 @@ fn writeAtoms(self: *MachO) !void {
     const gpa = self.base.allocator;
     const slice = self.sections.slice();
 
-    for (slice.items(.first_atom_index)) |first_atom_index, sect_id| {
+    for (slice.items(.first_atom_index), 0..) |first_atom_index, sect_id| {
         const header = slice.items(.header)[sect_id];
         var atom_index = first_atom_index;
 
@@ -2093,7 +2093,7 @@ fn calcSectionSizes(self: *MachO) !void {
     defer tracy.end();
 
     const slice = self.sections.slice();
-    for (slice.items(.header)) |*header, sect_id| {
+    for (slice.items(.header), 0..) |*header, sect_id| {
         if (header.size == 0) continue;
         if (self.requiresThunks()) {
             if (header.isCode() and !(header.type() == macho.S_SYMBOL_STUBS) and !mem.eql(u8, header.sectName(), "__stub_helper")) continue;
@@ -2124,7 +2124,7 @@ fn calcSectionSizes(self: *MachO) !void {
     }
 
     if (self.requiresThunks()) {
-        for (slice.items(.header)) |header, sect_id| {
+        for (slice.items(.header), 0..) |header, sect_id| {
             if (!header.isCode()) continue;
             if (header.type() == macho.S_SYMBOL_STUBS) continue;
             if (mem.eql(u8, header.sectName(), "__stub_helper")) continue;
@@ -2139,7 +2139,7 @@ fn allocateSegments(self: *MachO) !void {
     const tracy = trace(@src());
     defer tracy.end();
 
-    for (self.segments.items) |*segment, segment_index| {
+    for (self.segments.items, 0..) |*segment, segment_index| {
         const is_text_segment = mem.eql(u8, segment.segName(), "__TEXT");
         const base_size = if (is_text_segment) try load_commands.calcMinHeaderPad(self.base.allocator, &self.options, .{
             .segments = self.segments.items,
@@ -2205,7 +2205,7 @@ fn allocateSegment(self: *MachO, segment_index: u8, init_size: u64) !void {
     var start = init_size;
 
     const slice = self.sections.slice();
-    for (slice.items(.header)[indexes.start..indexes.end]) |*header, sect_id| {
+    for (slice.items(.header)[indexes.start..indexes.end], 0..) |*header, sect_id| {
         const alignment = try math.powi(u32, 2, header.@"align");
         const start_aligned = mem.alignForwardGeneric(u64, start, alignment);
         const n_sect = @intCast(u8, indexes.start + sect_id + 1);
@@ -2349,7 +2349,7 @@ fn getSectionPrecedence(header: macho.section_64) u8 {
 }
 
 fn writeSegmentHeaders(self: *MachO, writer: anytype) !void {
-    for (self.segments.items) |seg, i| {
+    for (self.segments.items, 0..) |seg, i| {
         const indexes = self.getSectionIndexes(@intCast(u8, i));
         var out_seg = seg;
         out_seg.cmdsize = @sizeOf(macho.segment_command_64);
@@ -2455,7 +2455,7 @@ fn collectRebaseData(self: *MachO, rebase: *Rebase) !void {
     }
 
     // Finally, unpack the rest.
-    for (slice.items(.header)) |header, sect_id| {
+    for (slice.items(.header), 0..) |header, sect_id| {
         switch (header.type()) {
             macho.S_LITERAL_POINTERS,
             macho.S_REGULAR,
@@ -2590,7 +2590,7 @@ fn collectBindData(self: *MachO, bind: *Bind) !void {
 
     // Finally, unpack the rest.
     const slice = self.sections.slice();
-    for (slice.items(.header)) |header, sect_id| {
+    for (slice.items(.header), 0..) |header, sect_id| {
         switch (header.type()) {
             macho.S_LITERAL_POINTERS,
             macho.S_REGULAR,
@@ -3301,7 +3301,7 @@ fn writeUuid(self: *MachO, args: struct {
         const amt = try self.base.file.preadAll(locals_buf, self.symtab_cmd.symoff);
         if (amt != locals_buf.len) return error.InputOutput;
 
-        const istab: usize = for (locals) |local, i| {
+        const istab: usize = for (locals, 0..) |local, i| {
             if (local.stab()) break i;
         } else locals.len;
         const nstabs = locals.len - istab;
@@ -3485,7 +3485,7 @@ pub fn getAtom(self: MachO, atom_index: AtomIndex) Atom {
 }
 
 fn getSegmentByName(self: MachO, segname: []const u8) ?u8 {
-    for (self.segments.items) |seg, i| {
+    for (self.segments.items, 0..) |seg, i| {
         if (mem.eql(u8, segname, seg.segName())) return @intCast(u8, i);
     } else return null;
 }
@@ -3509,7 +3509,7 @@ pub fn getLinkeditSegmentPtr(self: *MachO) *macho.segment_command_64 {
 
 pub fn getSectionByName(self: MachO, segname: []const u8, sectname: []const u8) ?u8 {
     // TODO investigate caching with a hashmap
-    for (self.sections.items(.header)) |header, i| {
+    for (self.sections.items(.header), 0..) |header, i| {
         if (mem.eql(u8, header.segName(), segname) and mem.eql(u8, header.sectName(), sectname))
             return @intCast(u8, i);
     } else return null;
@@ -3517,7 +3517,7 @@ pub fn getSectionByName(self: MachO, segname: []const u8, sectname: []const u8) 
 
 pub fn getSectionIndexes(self: MachO, segment_index: u8) struct { start: u8, end: u8 } {
     var start: u8 = 0;
-    const nsects = for (self.segments.items) |seg, i| {
+    const nsects = for (self.segments.items, 0..) |seg, i| {
         if (i == segment_index) break @intCast(u8, seg.nsects);
         start += @intCast(u8, seg.nsects);
     } else 0;
@@ -3840,7 +3840,7 @@ fn generateSymbolStabsForSymbol(
 
 fn logSegments(self: *MachO) void {
     log.debug("segments:", .{});
-    for (self.segments.items) |segment, i| {
+    for (self.segments.items, 0..) |segment, i| {
         log.debug("  segment({d}): {s} @{x} ({x}), sizeof({x})", .{
             i,
             segment.segName(),
@@ -3853,7 +3853,7 @@ fn logSegments(self: *MachO) void {
 
 fn logSections(self: *MachO) void {
     log.debug("sections:", .{});
-    for (self.sections.items(.header)) |header, i| {
+    for (self.sections.items(.header), 0..) |header, i| {
         log.debug("  sect({d}): {s},{s} @{x} ({x}), sizeof({x})", .{
             i + 1,
             header.segName(),
@@ -3891,9 +3891,9 @@ fn logSymtab(self: *MachO) void {
     const scoped_log = std.log.scoped(.symtab);
 
     scoped_log.debug("locals:", .{});
-    for (self.objects.items) |object, id| {
+    for (self.objects.items, 0..) |object, id| {
         scoped_log.debug("  object({d}): {s}", .{ id, object.name });
-        for (object.symtab) |sym, sym_id| {
+        for (object.symtab, 0..) |sym, sym_id| {
             if (object.in_symtab == null) continue;
             mem.set(u8, &buf, '_');
             scoped_log.debug("    %{d}: {s} @{x} in sect({d}), {s}", .{
@@ -3907,7 +3907,7 @@ fn logSymtab(self: *MachO) void {
     }
 
     scoped_log.debug("  object(-1)", .{});
-    for (self.locals.items) |sym, sym_id| {
+    for (self.locals.items, 0..) |sym, sym_id| {
         if (sym.undf()) continue;
         scoped_log.debug("    %{d}: {s} @{x} in sect({d}), {s}", .{
             sym_id,
@@ -3919,7 +3919,7 @@ fn logSymtab(self: *MachO) void {
     }
 
     scoped_log.debug("exports:", .{});
-    for (self.globals.items) |global, i| {
+    for (self.globals.items, 0..) |global, i| {
         const sym = self.getSymbol(global);
         if (sym.undf()) continue;
         if (sym.n_desc == N_DEAD) continue;
@@ -3934,7 +3934,7 @@ fn logSymtab(self: *MachO) void {
     }
 
     scoped_log.debug("imports:", .{});
-    for (self.globals.items) |global, i| {
+    for (self.globals.items, 0..) |global, i| {
         const sym = self.getSymbol(global);
         if (!sym.undf()) continue;
         if (sym.n_desc == N_DEAD) continue;
@@ -3949,7 +3949,7 @@ fn logSymtab(self: *MachO) void {
     }
 
     scoped_log.debug("GOT entries:", .{});
-    for (self.got_entries.items) |entry, i| {
+    for (self.got_entries.items, 0..) |entry, i| {
         const atom_sym = entry.getAtomSymbol(self);
         const target_sym = entry.getTargetSymbol(self);
         const target_sym_name = entry.getTargetSymbolName(self);
@@ -3971,7 +3971,7 @@ fn logSymtab(self: *MachO) void {
     }
 
     scoped_log.debug("__thread_ptrs entries:", .{});
-    for (self.tlv_ptr_entries.items) |entry, i| {
+    for (self.tlv_ptr_entries.items, 0..) |entry, i| {
         const atom_sym = entry.getAtomSymbol(self);
         const target_sym = entry.getTargetSymbol(self);
         const target_sym_name = entry.getTargetSymbolName(self);
@@ -3984,7 +3984,7 @@ fn logSymtab(self: *MachO) void {
     }
 
     scoped_log.debug("stubs entries:", .{});
-    for (self.stubs.items) |entry, i| {
+    for (self.stubs.items, 0..) |entry, i| {
         const atom_sym = entry.getAtomSymbol(self);
         const target_sym = entry.getTargetSymbol(self);
         const target_sym_name = entry.getTargetSymbolName(self);
@@ -3997,9 +3997,9 @@ fn logSymtab(self: *MachO) void {
     }
 
     scoped_log.debug("thunks:", .{});
-    for (self.thunks.items) |thunk, i| {
+    for (self.thunks.items, 0..) |thunk, i| {
         scoped_log.debug("  thunk({d})", .{i});
-        for (thunk.lookup.keys()) |target, j| {
+        for (thunk.lookup.keys(), 0..) |target, j| {
             const target_sym = self.getSymbol(target);
             const atom = self.getAtom(thunk.lookup.get(target).?);
             const atom_sym = self.getSymbol(atom.getSymbolWithLoc());
@@ -4016,7 +4016,7 @@ fn logSymtab(self: *MachO) void {
 fn logAtoms(self: *MachO) void {
     log.debug("atoms:", .{});
     const slice = self.sections.slice();
-    for (slice.items(.first_atom_index)) |first_atom_index, sect_id| {
+    for (slice.items(.first_atom_index), 0..) |first_atom_index, sect_id| {
         var atom_index = first_atom_index;
         if (atom_index == 0) continue;
 

--- a/src/MachO/Dylib.zig
+++ b/src/MachO/Dylib.zig
@@ -351,7 +351,7 @@ pub fn parseFromStub(
     var matcher = try TargetMatcher.init(allocator, target);
     defer matcher.deinit();
 
-    for (lib_stub.inner) |elem, stub_index| {
+    for (lib_stub.inner, 0..) |elem, stub_index| {
         const is_match = switch (elem) {
             .v3 => |stub| matcher.matchesArch(stub.archs),
             .v4 => |stub| matcher.matchesTarget(stub.targets),

--- a/src/MachO/Object.zig
+++ b/src/MachO/Object.zig
@@ -199,7 +199,7 @@ pub fn parse(self: *Object, allocator: Allocator, cpu_arch: std.Target.Cpu.Arch)
     var sorted_all_syms = try std.ArrayList(SymbolAtIndex).initCapacity(allocator, self.in_symtab.?.len);
     defer sorted_all_syms.deinit();
 
-    for (self.in_symtab.?) |_, index| {
+    for (self.in_symtab.?, 0..) |_, index| {
         sorted_all_syms.appendAssumeCapacity(.{ .index = @intCast(u32, index) });
     }
 
@@ -209,7 +209,7 @@ pub fn parse(self: *Object, allocator: Allocator, cpu_arch: std.Target.Cpu.Arch)
     // is kind enough to specify the symbols in the correct order.
     sort.sort(SymbolAtIndex, sorted_all_syms.items, self, SymbolAtIndex.lessThan);
 
-    for (sorted_all_syms.items) |sym_id, i| {
+    for (sorted_all_syms.items, 0..) |sym_id, i| {
         const sym = sym_id.getSymbol(self);
 
         if (sym.sect() and self.source_section_index_lookup[sym.n_sect - 1] == -1) {
@@ -378,7 +378,7 @@ pub fn splitRegularSections(self: *Object, macho_file: *MachO, object_id: u32) !
     const gpa = macho_file.base.allocator;
 
     const sections = self.getSourceSections();
-    for (sections) |sect, id| {
+    for (sections, 0..) |sect, id| {
         if (sect.isDebug()) continue;
         const out_sect_id = (try macho_file.getOutputSection(sect)) orelse {
             log.debug("  unhandled section '{s},{s}'", .{ sect.segName(), sect.sectName() });
@@ -398,7 +398,7 @@ pub fn splitRegularSections(self: *Object, macho_file: *MachO, object_id: u32) !
     }
 
     if (self.in_symtab == null) {
-        for (sections) |sect, id| {
+        for (sections, 0..) |sect, id| {
             if (sect.isDebug()) continue;
             const out_sect_id = (try macho_file.getOutputSection(sect)) orelse continue;
             if (sect.size == 0) continue;
@@ -444,7 +444,7 @@ pub fn splitRegularSections(self: *Object, macho_file: *MachO, object_id: u32) !
     var sorted_sections = try gpa.alloc(SortedSection, sections.len);
     defer gpa.free(sorted_sections);
 
-    for (sections) |sect, id| {
+    for (sections, 0..) |sect, id| {
         sorted_sections[id] = .{ .header = sect, .id = @intCast(u8, id) };
     }
 
@@ -802,7 +802,7 @@ fn parseUnwindInfo(self: *Object, macho_file: *MachO, object_id: u32) !void {
     try self.parseRelocs(gpa, sect_id);
     const relocs = self.getRelocs(sect_id);
 
-    for (unwind_records) |record, record_id| {
+    for (unwind_records, 0..) |record, record_id| {
         const offset = record_id * @sizeOf(macho.compact_unwind_entry);
         const rel_pos = filterRelocs(
             relocs,
@@ -855,7 +855,7 @@ pub fn getSourceSectionByName(self: Object, segname: []const u8, sectname: []con
 
 pub fn getSourceSectionIndexByName(self: Object, segname: []const u8, sectname: []const u8) ?u8 {
     const sections = self.getSourceSections();
-    for (sections) |sect, i| {
+    for (sections, 0..) |sect, i| {
         if (mem.eql(u8, segname, sect.segName()) and mem.eql(u8, sectname, sect.sectName()))
             return @intCast(u8, i);
     } else return null;

--- a/src/MachO/UnwindInfo.zig
+++ b/src/MachO/UnwindInfo.zig
@@ -123,7 +123,7 @@ const Page = struct {
             ctx.page.start + ctx.page.count,
         });
         try writer.print("  encodings (count = {d})\n", .{ctx.page.page_encodings_count});
-        for (ctx.page.page_encodings[0..ctx.page.page_encodings_count]) |record_id, i| {
+        for (ctx.page.page_encodings[0..ctx.page.page_encodings_count], 0..) |record_id, i| {
             const record = ctx.info.records.items[record_id];
             const enc = record.compactUnwindEncoding;
             try writer.print("    {d}: 0x{x:0>8}\n", .{ ctx.info.common_encodings_count + i, enc });
@@ -202,7 +202,7 @@ pub fn scanRelocs(macho_file: *MachO) !void {
     if (macho_file.getSectionByName("__TEXT", "__unwind_info") == null) return;
 
     const cpu_arch = macho_file.options.target.cpu_arch.?;
-    for (macho_file.objects.items) |*object, object_id| {
+    for (macho_file.objects.items, 0..) |*object, object_id| {
         const unwind_records = object.getUnwindRecords();
         for (object.exec_atoms.items) |atom_index| {
             const record_id = object.unwind_records_lookup.get(atom_index) orelse continue;
@@ -241,7 +241,7 @@ pub fn collect(info: *UnwindInfo, macho_file: *MachO) !void {
     defer atom_indexes.deinit();
 
     // TODO handle dead stripping
-    for (macho_file.objects.items) |*object, object_id| {
+    for (macho_file.objects.items, 0..) |*object, object_id| {
         log.debug("collecting unwind records in {s} ({d})", .{ object.name, object_id });
         const unwind_records = object.getUnwindRecords();
 
@@ -332,7 +332,7 @@ pub fn collect(info: *UnwindInfo, macho_file: *MachO) !void {
     try info.records_lookup.ensureTotalCapacity(info.gpa, @intCast(u32, atom_indexes.items.len));
 
     var maybe_prev: ?macho.compact_unwind_entry = null;
-    for (records.items) |record, i| {
+    for (records.items, 0..) |record, i| {
         const record_id = blk: {
             if (maybe_prev) |prev| {
                 const is_dwarf = UnwindEncoding.isDwarf(record.compactUnwindEncoding, cpu_arch);
@@ -480,7 +480,7 @@ pub fn collect(info: *UnwindInfo, macho_file: *MachO) !void {
 
     // Save indices of records requiring LSDA relocation
     try info.lsdas_lookup.ensureTotalCapacity(info.gpa, @intCast(u32, info.records.items.len));
-    for (info.records.items) |rec, i| {
+    for (info.records.items, 0..) |rec, i| {
         info.lsdas_lookup.putAssumeCapacityNoClobber(@intCast(RecordIndex, i), @intCast(u32, info.lsdas.items.len));
         if (rec.lsda == 0) continue;
         try info.lsdas.append(info.gpa, @intCast(RecordIndex, i));
@@ -553,7 +553,7 @@ pub fn write(info: *UnwindInfo, macho_file: *MachO) !void {
     const cpu_arch = macho_file.options.target.cpu_arch.?;
 
     log.debug("Personalities:", .{});
-    for (info.personalities[0..info.personalities_count]) |target, i| {
+    for (info.personalities[0..info.personalities_count], 0..) |target, i| {
         const atom_index = macho_file.getGotAtomIndexForSymbol(target).?;
         const atom = macho_file.getAtom(atom_index);
         const sym = macho_file.getSymbol(atom.getSymbolWithLoc());
@@ -577,7 +577,7 @@ pub fn write(info: *UnwindInfo, macho_file: *MachO) !void {
         }
     }
 
-    for (info.records.items) |record, i| {
+    for (info.records.items, 0..) |record, i| {
         log.debug("Unwind record at offset 0x{x}", .{i * @sizeOf(macho.compact_unwind_entry)});
         log.debug("  start: 0x{x}", .{record.rangeStart});
         log.debug("  length: 0x{x}", .{record.rangeLength});
@@ -617,7 +617,7 @@ pub fn write(info: *UnwindInfo, macho_file: *MachO) !void {
     const pages_base_offset = @intCast(u32, size - (info.pages.items.len * second_level_page_bytes));
     const lsda_base_offset = @intCast(u32, pages_base_offset -
         (info.lsdas.items.len * @sizeOf(macho.unwind_info_section_header_lsda_index_entry)));
-    for (info.pages.items) |page, i| {
+    for (info.pages.items, 0..) |page, i| {
         assert(page.count > 0);
         const first_entry = info.records.items[page.start];
         try writer.writeStruct(macho.unwind_info_section_header_index_entry{

--- a/src/MachO/dead_strip.zig
+++ b/src/MachO/dead_strip.zig
@@ -263,7 +263,7 @@ fn mark(macho_file: *MachO, roots: AtomTable, alive: *AtomTable) !void {
         }
     }
 
-    for (macho_file.objects.items) |_, object_id| {
+    for (macho_file.objects.items, 0..) |_, object_id| {
         // Traverse unwind and eh_frame records noting if the source symbol has been marked, and if so,
         // marking all references as live.
         try markUnwindRecords(macho_file, @intCast(u32, object_id), alive);

--- a/src/MachO/dyld_info/Rebase.zig
+++ b/src/MachO/dyld_info/Rebase.zig
@@ -45,7 +45,7 @@ pub fn finalize(rebase: *Rebase, gpa: Allocator) !void {
 
     var start: usize = 0;
     var seg_id: ?u8 = null;
-    for (rebase.entries.items) |entry, i| {
+    for (rebase.entries.items, 0..) |entry, i| {
         if (seg_id != null and seg_id.? == entry.segment_id) continue;
         try finalizeSegment(rebase.entries.items[start..i], writer);
         seg_id = entry.segment_id;

--- a/src/MachO/dyld_info/bind.zig
+++ b/src/MachO/dyld_info/bind.zig
@@ -51,7 +51,7 @@ pub fn Bind(comptime Ctx: type, comptime Target: type) type {
 
             var start: usize = 0;
             var seg_id: ?u8 = null;
-            for (self.entries.items) |entry, i| {
+            for (self.entries.items, 0..) |entry, i| {
                 if (seg_id != null and seg_id.? == entry.segment_id) continue;
                 try finalizeSegment(self.entries.items[start..i], ctx, writer);
                 seg_id = entry.segment_id;

--- a/src/MachO/eh_frame.zig
+++ b/src/MachO/eh_frame.zig
@@ -15,7 +15,7 @@ const UnwindInfo = @import("UnwindInfo.zig");
 pub fn scanRelocs(macho_file: *MachO) !void {
     const gpa = macho_file.base.allocator;
 
-    for (macho_file.objects.items) |*object, object_id| {
+    for (macho_file.objects.items, 0..) |*object, object_id| {
         var cies = std.AutoHashMap(u32, void).init(gpa);
         defer cies.deinit();
 
@@ -107,7 +107,7 @@ pub fn write(macho_file: *MachO, unwind_info: *UnwindInfo) !void {
 
     var eh_frame_offset: u32 = 0;
 
-    for (macho_file.objects.items) |*object, object_id| {
+    for (macho_file.objects.items, 0..) |*object, object_id| {
         try eh_records.ensureUnusedCapacity(2 * @intCast(u32, object.exec_atoms.items.len));
 
         var cies = std.AutoHashMap(u32, u32).init(gpa);
@@ -406,7 +406,7 @@ pub fn EhFrameRecord(comptime is_mutable: bool) type {
             var creader = std.io.countingReader(stream.reader());
             const reader = creader.reader();
 
-            for (aug_str) |ch, i| switch (ch) {
+            for (aug_str, 0..) |ch, i| switch (ch) {
                 'z' => if (i > 0) {
                     return error.MalformedAugmentationString;
                 } else {
@@ -466,7 +466,7 @@ pub fn EhFrameRecord(comptime is_mutable: bool) type {
             var creader = std.io.countingReader(stream.reader());
             const reader = creader.reader();
 
-            for (aug_str) |ch, i| switch (ch) {
+            for (aug_str, 0..) |ch, i| switch (ch) {
                 'z' => if (i > 0) {
                     return error.MalformedAugmentationString;
                 } else {

--- a/src/MachO/thunks.zig
+++ b/src/MachO/thunks.zig
@@ -321,7 +321,7 @@ fn createThunkAtom(macho_file: *MachO) !AtomIndex {
 fn getThunkIndex(macho_file: *MachO, atom_index: AtomIndex) ?ThunkIndex {
     const atom = macho_file.getAtom(atom_index);
     const sym = macho_file.getSymbol(atom.getSymbolWithLoc());
-    for (macho_file.thunks.items) |thunk, i| {
+    for (macho_file.thunks.items, 0..) |thunk, i| {
         if (thunk.len == 0) continue;
 
         const thunk_atom_index = thunk.getStartAtomIndex();

--- a/src/Wasm.zig
+++ b/src/Wasm.zig
@@ -415,14 +415,14 @@ pub fn dataCount(wasm: Wasm) u32 {
 pub fn flush(wasm: *Wasm) !void {
     try wasm.parsePositionals(wasm.options.positionals);
     try wasm.setupLinkerSymbols();
-    for (wasm.objects.items) |_, obj_idx| {
+    for (wasm.objects.items, 0..) |_, obj_idx| {
         try wasm.resolveSymbolsInObject(@intCast(u16, obj_idx));
     }
     try wasm.resolveSymbolsInArchives();
     try wasm.resolveLazySymbols();
     try wasm.setupInitFunctions();
     try wasm.checkUndefinedSymbols();
-    for (wasm.objects.items) |*object, obj_idx| {
+    for (wasm.objects.items, 0..) |*object, obj_idx| {
         try object.parseIntoAtoms(@intCast(u16, obj_idx), wasm);
     }
     try wasm.validateFeatures();
@@ -516,7 +516,7 @@ fn resolveSymbolsInObject(wasm: *Wasm, object_index: u16) !void {
     const object: Object = wasm.objects.items[object_index];
     log.debug("Resolving symbols in object: '{s}'", .{object.name});
 
-    for (object.symtable) |symbol, i| {
+    for (object.symtable, 0..) |symbol, i| {
         const sym_index = @intCast(u32, i);
         const location: SymbolWithLoc = .{
             .file = object_index,
@@ -1059,7 +1059,7 @@ fn checkUndefinedSymbols(wasm: *const Wasm) !void {
 /// After all functions have been inserted, the functions will be ordered based
 /// on their priority.
 fn setupInitFunctions(wasm: *Wasm) !void {
-    for (wasm.objects.items) |object, file_index| {
+    for (wasm.objects.items, 0..) |object, file_index| {
         try wasm.init_funcs.ensureUnusedCapacity(wasm.base.allocator, object.init_funcs.len);
         for (object.init_funcs) |init_func| {
             const symbol = object.symtable[init_func.symbol_index];
@@ -1275,7 +1275,7 @@ fn setupTLSRelocationsFunction(wasm: *Wasm) !void {
 
     // locals (we have none)
     try writer.writeByte(0);
-    for (wasm.globals.got_symbols.items) |got_loc, got_index| {
+    for (wasm.globals.got_symbols.items, 0..) |got_loc, got_index| {
         const sym: *Symbol = got_loc.getSymbol(wasm);
         if (!sym.isTLS()) continue; // only relocate TLS symbols
         if (sym.tag == .data and sym.isDefined()) {
@@ -1837,7 +1837,7 @@ fn validateFeatures(wasm: *Wasm) !void {
 
     // extract all the used, disallowed and required features from each
     // linked object file so we can test them.
-    for (wasm.objects.items) |object, object_index| {
+    for (wasm.objects.items, 0..) |object, object_index| {
         for (object.features) |feature| {
             const value = @intCast(u16, object_index) << 1 | @as(u1, 1);
             switch (feature.prefix) {
@@ -1864,7 +1864,7 @@ fn validateFeatures(wasm: *Wasm) !void {
     // when we infer the features, we allow each feature found in the 'used' set
     // and insert it into the 'allowed' set. When features are not inferred,
     // we validate that a used feature is allowed.
-    for (used) |used_set, used_index| {
+    for (used, 0..) |used_set, used_index| {
         const is_enabled = @truncate(u1, used_set) != 0;
         if (!is_enabled) continue;
         const feature = @intToEnum(types.Feature.Tag, used_index);
@@ -1924,7 +1924,7 @@ fn validateFeatures(wasm: *Wasm) !void {
         }
 
         // validate the linked object file has each required feature
-        for (required) |required_feature, feature_index| {
+        for (required, 0..) |required_feature, feature_index| {
             const is_required = @truncate(u1, required_feature) != 0;
             if (is_required and !object_used_features[feature_index]) {
                 log.err("feature '{}' is required but not used in linked object", .{(@intToEnum(types.Feature.Tag, feature_index))});

--- a/src/Wasm/Object.zig
+++ b/src/Wasm/Object.zig
@@ -908,7 +908,7 @@ pub fn parseIntoAtoms(object: *Object, object_index: u16, wasm_bin: *Wasm) !void
         list.deinit();
     } else symbol_for_segment.deinit();
 
-    for (object.symtable) |symbol, symbol_index| {
+    for (object.symtable, 0..) |symbol, symbol_index| {
         switch (symbol.tag) {
             .function, .data, .section => if (!symbol.isUndefined()) {
                 const gop = try symbol_for_segment.getOrPut(.{ .kind = symbol.tag, .index = symbol.index });
@@ -922,7 +922,7 @@ pub fn parseIntoAtoms(object: *Object, object_index: u16, wasm_bin: *Wasm) !void
         }
     }
 
-    for (object.relocatable_data) |relocatable_data, index| {
+    for (object.relocatable_data, 0..) |relocatable_data, index| {
         const final_index = (try wasm_bin.getMatchingSegment(wasm_bin.base.allocator, object_index, @intCast(u32, index))) orelse {
             continue; // found unknown section, so skip parsing into atom as we do not know how to handle it.
         };

--- a/src/Wasm/emit_wasm.zig
+++ b/src/Wasm/emit_wasm.zig
@@ -274,7 +274,7 @@ fn emitDataNamesSection(wasm: *Wasm, gpa: std.mem.Allocator, writer: anytype) !v
     const sec_writer = section_list.writer();
 
     try leb.writeULEB128(sec_writer, wasm.dataCount());
-    for (wasm.data_segments.keys()) |key, index| {
+    for (wasm.data_segments.keys(), 0..) |key, index| {
         if (std.mem.eql(u8, key, ".bss") and !wasm.options.import_memory) continue;
         try leb.writeULEB128(sec_writer, @intCast(u32, index));
         try leb.writeULEB128(sec_writer, @intCast(u32, key.len));

--- a/src/Wasm/sections.zig
+++ b/src/Wasm/sections.zig
@@ -286,7 +286,7 @@ pub const Types = struct {
     /// the type section. When it does exist, it will return its index
     /// otherwise, returns `null`.
     pub fn find(self: Types, func_type: std.wasm.Type) ?u32 {
-        return for (self.items.items) |ty, index| {
+        return for (self.items.items, 0..) |ty, index| {
             if (ty.eql(func_type)) {
                 return @intCast(u32, index);
             }

--- a/src/tapi.zig
+++ b/src/tapi.zig
@@ -124,7 +124,7 @@ pub const LibStub = struct {
                 log.debug("trying to parse as []TbdV4", .{});
                 const inner = lib_stub.yaml.parse([]TbdV4) catch break :err;
                 var out = try lib_stub.yaml.arena.allocator().alloc(Tbd, inner.len);
-                for (inner) |doc, i| {
+                for (inner, 0..) |doc, i| {
                     out[i] = .{ .v4 = doc };
                 }
                 break :blk out;
@@ -142,7 +142,7 @@ pub const LibStub = struct {
                 log.debug("trying to parse as []TbdV3", .{});
                 const inner = lib_stub.yaml.parse([]TbdV3) catch break :err;
                 var out = try lib_stub.yaml.arena.allocator().alloc(Tbd, inner.len);
-                for (inner) |doc, i| {
+                for (inner, 0..) |doc, i| {
                     out[i] = .{ .v3 = doc };
                 }
                 break :blk out;

--- a/src/tapi/yaml.zig
+++ b/src/tapi/yaml.zig
@@ -84,7 +84,7 @@ pub const Value = union(ValueType) {
 
                 const first = list[0];
                 if (first.is_compound()) {
-                    for (list) |elem, i| {
+                    for (list, 0..) |elem, i| {
                         try writer.writeByteNTimes(' ', args.indentation);
                         try writer.writeAll("- ");
                         try elem.stringify(writer, .{
@@ -99,7 +99,7 @@ pub const Value = union(ValueType) {
                 }
 
                 try writer.writeAll("[ ");
-                for (list) |elem, i| {
+                for (list, 0..) |elem, i| {
                     try elem.stringify(writer, args);
                     if (i < len - 1) {
                         try writer.writeAll(", ");
@@ -112,7 +112,7 @@ pub const Value = union(ValueType) {
                 const len = keys.len;
                 if (len == 0) return;
 
-                for (keys) |key, i| {
+                for (keys, 0..) |key, i| {
                     if (!args.should_inline_first_key or i != 0) {
                         try writer.writeByteNTimes(' ', args.indentation);
                     }
@@ -292,7 +292,7 @@ pub const Yaml = struct {
         switch (@typeInfo(T)) {
             .Array => |info| {
                 var parsed: T = undefined;
-                for (self.docs.items) |doc, i| {
+                for (self.docs.items, 0..) |doc, i| {
                     parsed[i] = try self.parseValue(info.child, doc);
                 }
                 return parsed;
@@ -301,7 +301,7 @@ pub const Yaml = struct {
                 switch (info.size) {
                     .Slice => {
                         var parsed = try self.arena.allocator().alloc(info.child, self.docs.items.len);
-                        for (self.docs.items) |doc, i| {
+                        for (self.docs.items, 0..) |doc, i| {
                             parsed[i] = try self.parseValue(info.child, doc);
                         }
                         return parsed;
@@ -393,7 +393,7 @@ pub const Yaml = struct {
                 }
 
                 var parsed = try arena.alloc(ptr_info.child, value.list.len);
-                for (value.list) |elem, i| {
+                for (value.list, 0..) |elem, i| {
                     parsed[i] = try self.parseValue(ptr_info.child, elem);
                 }
                 return parsed;
@@ -407,7 +407,7 @@ pub const Yaml = struct {
         if (array_info.len != list.len) return error.ArraySizeMismatch;
 
         var parsed: T = undefined;
-        for (list) |elem, i| {
+        for (list, 0..) |elem, i| {
             parsed[i] = try self.parseValue(array_info.child, elem);
         }
 

--- a/src/test.zig
+++ b/src/test.zig
@@ -24,7 +24,7 @@ test "end-to-end" {
     var ctx = TestContext.init();
     defer ctx.deinit();
 
-    try @import("end_to_end_tests").addCases(&ctx);
+    try @import("../test/test.zig").addCases(&ctx);
     try ctx.run();
 }
 

--- a/test/test.zig
+++ b/test/test.zig
@@ -1,6 +1,6 @@
 const std = @import("std");
 const CrossTarget = std.zig.CrossTarget;
-const TestContext = @import("test_base").TestContext;
+const TestContext = @import("../src/test.zig").TestContext;
 
 const linux_x86_64 = CrossTarget{
     .cpu_arch = .x86_64,

